datapage/Datapage.ts

```diff
@@ -0,0 +1,151 @@
+                        message: err.message,
+                    },
+                ],
+            }
+        }
+    }
+
+    // Validate the datapage JSON file against the DataPageJson schema
+    if (Value.Check(DataPageJsonTypeObject, datapageJson)) {
+        return { datapageJson, parseErrors: [] }
+    } else {
+        return {
+            datapageJson,
+            parseErrors: [
+                ...Value.Errors(DataPageJsonTypeObject, datapageJson),
+            ].map(({ message, path }) => ({ message, path })),
+        }
+    }
+}
+
+/**
+ * Get the datapage companion gdoc, if any.
+ *
+ * When previewing, we want to render the datapage from the live gdoc.
+ * Otherwise, we're baking from the gdoc parsed and saved in the database
+ * following a visit to /admin/gdocs/[googleDocId]/preview
+ */
+export const getDatapageGdoc = async (
+    datapageJson: DataPageJson,
+    isPreviewing: boolean,
+    publishedExplorersBySlug?: Record<string, ExplorerProgram>
+): Promise<DataPageGdoc | null> => {
+    // Get the google doc id from the datapage JSON file and return early if
+    // none found
+    const googleDocId =
+        getLinkType(datapageJson.googleDocEditLink) === "gdoc"
+            ? getUrlTarget(datapageJson.googleDocEditLink)
+            : null
+
+    if (!googleDocId) return null
+
+    // When previewing, we want to render the datapage from the live gdoc,
+    // but only if the user has set up the necessary auth keys to access the
+    // Google Doc API. This won't be the case for external contributors or
+    // possibly data engineers focusing on the data pipeline. In those
+    // cases, we grab the gdoc found in the database, if any.
+    const rawDatapageGdoc =
+        isPreviewing && publishedExplorersBySlug && Gdoc.areGdocAuthKeysSet()
+            ? await Gdoc.getGdocFromContentSource(
+                  googleDocId,
+                  publishedExplorersBySlug,
+                  GdocsContentSource.Gdocs
+              )
+            : await Gdoc.findOneBy({ id: googleDocId })
+
+    if (!rawDatapageGdoc) return null
+
+    // Split the gdoc content into fields using the heading one texts as field
+    // names. Any top level key that is not in AllowedDataPageGdocFields will
+    // be silently filtered out. Beyond that, any allowed field can contain the
+    // whole range of OwidEnrichedGdocBlock types.
+    return parseGdocContentFromAllowedLevelOneHeadings(rawDatapageGdoc)
+}
+
+/*
+ * Takes a gdoc and splits its content into sections based on the heading 1s.
+
+* The heading 1 texts are used as keys, which can represent a nested structure,
+ * e.g. `descriptionFromSource.content`. 
+ *
+ * Validation: only a subset of the possible fields found is allowed (see
+ * AllowedDataPageGdocFields). This means the gdoc can contain extra heading
+ * one texts (e.g. for documentation) that will be ignored without raising
+ * errors.
+ *  */
+export const parseGdocContentFromAllowedLevelOneHeadings = (
```

Comment:
The implementation is simple enough, but I'm not super keen on this way of storing information in a google doc. It's like a new stateful metasyntax, which could lead to all sorts of parsing and migration issues.

What other options did you consider? Was there anything disqualifying about the following sort of approaches?
```
{.datapage}
title: Banana Production
[.+topicTagLinks]
CO2 and greenhouse gas emissions
Climate change
[]
{}
```

```
{.field}
id: title
[.+value]
Banana Production
[]
{}
```

Reply:
```
{.datapage}
title: Banana Production
[.+topicTagLinks]
CO2 and greenhouse gas emissions
Climate change
[]
{}
```
I discarded that one because it would lead to creating one block per field, with those blocks being effectively identical wrappers with no additional functionality other than rendering children. So quite a bit of code, and not much to show for.

```
{.field}
id: title
[.+value]
Banana Production
[]
{}
```
This is the one I considered as a solid alternative for the current way of doing things.

As [I pointed out previously](https://owid.slack.com/archives/C04L4BHSFSB/p1682093096695729?thread_ts=1682093006.813349&cid=C04L4BHSFSB), I'm simply not seeing very strong arguments to justify spending time refactoring right now. The whole document is still 100% ArchieML compliant so, an equally simple parser / splitter wouldn't be too much of an issue should we ever need to migrate.

So I'm not saying this is a bad idea, quite the opposite, but I'm just not convinced that paying the price now makes sense in light of the remaining work on data pages which has a visible impact for readers.

A final alternative would be to have one field per gdoc. [This approach was discarded for the prototype as it created too much authoring overhead](https://owid.slack.com/archives/C04L4BHSFSB/p1682093231292949?thread_ts=1682093006.813349&cid=C04L4BHSFSB). When we get to the point where this fields/fragments need to be reused across data pages, then we might reevaluate and provide a way to reference those fragments in the JSON / ETL, although making this the only way to have rich text in data pages might be overkill - so they'll likely co-exist.